### PR TITLE
HAWQ-932. Added pxf service address to curl resolution.

### PR DESCRIPTION
**Pull request by sansanichfb**

### CODE REVIEW

**Comment by GodenYao** [comment]
churl_init_upload and  churl_init_download only has 1 tag difference ( context->upload = true vs. false) probably worth refactoring to a single function and invoked by these 2, which will make it semantic readable as well as easy to maintain

**Comment by kavinderd** [comment]
LGTM. Agree with @GodenYao on refactoring out the common code

File: src/backend/access/external/libchurl.c [hunk 1]
```diff
@@ -312,6 +312,14 @@ CHURL_HANDLE churl_init_upload(const char* url, CHURL_HEADERS headers)
 	context->upload = true;
 	clear_error_buffer(context);
 
+	/* needed to resolve pxf service address */
+	struct curl_slist *resolve_hosts = NULL;
+	char *pxf_host_entry = (char *) palloc0(strlen(pxf_service_address) + strlen(LocalhostIpV4Entry) + 1);
```
Comment (kavinderd) [inline src/backend/access/external/libchurl.c]: Is `pxf_host_entry` pfree'd?

File: src/backend/access/external/libchurl.c [hunk 2]
```diff
@@ -312,6 +312,14 @@ CHURL_HANDLE churl_init_upload(const char* url, CHURL_HEADERS headers)
...
+	strcat(pxf_host_entry, pxf_service_address);
```
Comment (shivzone) [inline src/backend/access/external/libchurl.c]: This step might be unnecssary if the pxf_service_address itself is based on the IP address

Comment (sansanichfb) [inline src/backend/access/external/libchurl.c]: For case when user created external table referring to "localhost" it's needed.

Comment (shivzone) [inline src/backend/access/external/libchurl.c]: Yes. I was suggesting that we add the below OPT only when pxf_service_address is not based on IP address

File: src/backend/utils/adt/pxf_functions.c [hunk 3]
```diff
@@ -19,6 +19,7 @@
 
 #include "postgres.h"
 
+#include "access/hd_work_mgr.h"
```
Comment (sansanichfb) [inline src/backend/utils/adt/pxf_functions.c]: Without this fix I was getting:
```
pxf_functions.c:46:10: warning: implicit declaration of function 'get_pxf_item_metadata' is invalid in C99
      [-Wimplicit-function-declaration]
        items = get_pxf_item_metadata(profile_cstr, pattern_cstr, NULL);
                ^
pxf_functions.c:46:8: warning: incompatible integer to pointer conversion assigning to 'List *' (aka 'struct List *') from 'int'
      [-Wint-conversion]
        items = get_pxf_item_metadata(profile_cstr, pattern_cstr, NULL);
              ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

File: src/backend/utils/adt/pxf_functions.c [hunk 4]
```diff
@@ -43,7 +44,7 @@ pxf_item_fields_enum_start(text *profile, text *pattern)
 	char *profile_cstr = text_to_cstring(profile);
 	char *pattern_cstr = text_to_cstring(pattern);
 
-	items = get_pxf_item_metadata(profile_cstr, pattern_cstr, NULL);
+	items = get_pxf_item_metadata(profile_cstr, pattern_cstr, InvalidOid);
```
Comment (sansanichfb) [inline src/backend/utils/adt/pxf_functions.c]: One more warning.

File: src/backend/catalog/external/externalmd.c [hunk 5]
```diff
@@ -84,7 +84,7 @@ List *ParsePxfEntries(StringInfo json, char *profile, Oid dboid)
 	{
 		struct json_object *jsonItem = json_object_array_get_idx(jsonItems, i);
 		PxfItem *pxfItem = ParsePxfItem(jsonItem, profile);
-		if (dboid != NULL)
+		if (dboid != InvalidOid)
```
Comment (sansanichfb) [inline src/backend/catalog/external/externalmd.c]: Warning.

File: src/backend/access/external/pxfmasterapi.c [hunk 6]
```diff
@@ -434,5 +434,5 @@ List* get_and_cache_external_metadata(GPHDUri* hadoop_uri, char *profile, char *
 
 List* get_no_cache_external_metadata(GPHDUri* hadoop_uri, char *profile, char *pattern, ClientContext *client_context)
 {
-	return get_external_metadata(hadoop_uri, profile, pattern, client_context, NULL);
+	return get_external_metadata(hadoop_uri, profile, pattern, client_context, InvalidOid);
```
Comment (sansanichfb) [inline src/backend/access/external/pxfmasterapi.c]: Fixed warning.

**Comment by shivzone** [comment]
+1

**Pull request closed without merging**